gossip/handler.go

```diff
@@ -856,11 +872,51 @@ func (h *handler) handle(p *peer) error {
 	// after this will be sent via broadcasts.
 	h.syncTransactions(p, h.txpool.SampleHashes(h.config.Protocol.MaxInitialTxHashesSend))
 
-	// Handle incoming messages until the connection is torn down
+	// Handle incoming messages until the connection is torn down or the inactivity
+	// timer times out.
+	var noOfApplicationErrors = 0
 	for {
-		if err := h.handleMsg(p); err != nil {
-			p.Log().Debug("Message handling failed", "err", err)
-			return err
+		// progress and application
+		progressWatchDogTimer := time.NewTimer(noProgressTime)
+		applicationWatchDogTimer := time.NewTimer(noAppMessageTime)
+		select {
+		case <-progressWatchDogTimer.C:
+			// If self syncing, don't check peer progress
+			if !h.syncStatus.AcceptTxs() {
+				progressWatchDogTimer.Reset(noProgressTime)
+				break
+			}
+			if p.IsPeerProgressing() {
+				progressWatchDogTimer.Reset(noProgressTime)
+			} else {
+				p.Log().Warn("progress timer timeout: ", "name", p.Name(), "node", p.Node().String())
+				discfilter.Ban(p.ID())
+				return ErrorProgressTimeout
+			}
+		case <- applicationWatchDogTimer.C:
+			if p.IsApplicationProgressing() {
+				applicationWatchDogTimer.Reset(noAppMessageTime)
+			} else {
+				p.Log().Warn("application timer timeout: ", "name", p.Name(), "node", p.Node().String())
+				discfilter.Ban(p.ID())
+				return ErrorApplicationTimeout
+			}
+		default:
+			err := h.handleMsg(p)
+			if err != nil {
+				p.Log().Debug("Message handling failed", "err", err)
+				if strings.Contains(err.Error(), errorToString[ErrPeerNotProgressing]) {
```

Comment:
We can use `errors.Is()` only to compare errors. But in this place, the error is defined as a string.
If we want to change it, we should define all the errors as errors.New().

Reply:
Yes agreed. If there are more such string based errors instead of `errors.New()` based ones (which I believe would be better) - then this should go into a separate PR to address. So up to you if you want to do anything in this PR.